Kernel/Scheduler.cpp

```diff
@@ -105,6 +105,33 @@ Thread& Scheduler::pull_next_runnable_thread()
     return *Processor::idle_thread();
 }
 
+Thread* Scheduler::peek_next_runnable_thread()
```

Comment:
Would it make sense to use this new method in `pull_next_runnable_thread`?

Reply:
Nevermind, I just tried that myself and it's a major mess.

Reply:
The other advantage of keeping them separate is that we can convert it to a SharedSpinLock, which will be especially beneficiary on SMP. Then you can have multiple processors peek at the same time but only one pull at a time.

Reply:
Oh just realized I should probably cherry-pick those SharedSpinLocks from my SMP PR ;-)